keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,705 @@
+
+[The documentation of node addresses] states that "the usage of these
+fields varies depending on your cloud provider or bare metal
+configuration" but suggests that "External" addresses are "typically the
+IP address of the node that is externally routable (available from
+outside the cluster)" and "Internal" addresses are "typically the IP
+address of the node that is routable only within the cluster".
+
+Beyond that there is also somewhat of a consensus that "Internal" IPs
+route directly to the node, while "External" IPs may pass through some
+sort of address-rewriting infrastructure (and thus be less efficient
+for node-to-node communication). That is, "Internal" and "External"
+are understood to map directly to the "internal"/"private" vs
+"external"/"public" IPv4 addresses in most public cloud APIs.
+
+For IPs that don't clearly obey this distinction there is not currently
+consensus on how they should be labelled:
+
+  - Clouds that support IPv6 may provide addresses which are both
+    globally accessible and directly routed; calling them either
+    Internal or External risks having some components misinterpret their
+    semantics.
+
+  - The vSphere provider claims that all IPs that it knows about are
+    *both* "Internal" and "External".
+
+  - In bare-metal environments, nodes have only a single IP, and that IP
+    is always called "Internal".
+
+[The documentation of node addresses]: https://kubernetes.io/docs/concepts/architecture/nodes/#addresses
+
+#### Backward Compatibility
+
+`Node.Status.Addresses` needs to stick around as long as `v1.Node`
+does, but it can be essentially frozen with its current semantics.
+Existing cloud providers may continue to set `Node.Status.Addresses`
+with values that are not entirely equivalent to the `Node.Status.IPs`
+values, for backward compatibility. For new cloud providers we may be
+able to have `Node.Status.Addresses` be automatically generated from
+`Node.Status.IPs` and `Node.Status.Hostname` by
+cloud-controller-manager.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+```
+<<[UNRESOLVED inconsistency ]>>
+Security issues related to Node.Status.Addresses and Node.Status.IPs
+containing different values?
+<<[/UNRESOLVED]>>
+```
+
+## Design Details
+
+### Implementation
+
+#### Hostname
+
+With the deprecation of `Node.Status.Addresses`, this value will be
+moved to `Node.Status.Hostname`, but does not otherwise need to be
+changed.
+
+```
+<<[UNRESOLVED hostname-label ]>>
+Is it a bug that `Node.Status.Hostname`/`NodeHostName` and the
+`"kubernetes.io/hostname"` label don't necessarily match? (NB: This
+is an existing behavior.)
```

Comment:
OK, so maybe we should create `"kubernetes.io/nodename"` and deprecate `"kubernetes.io/hostname"`? (Except that we presumably can't actually get rid of "`"kubernetes.io/hostname"`, so this would just make things extra-confusing in the case where the actual hostname was not the node name.)